TO MERGE DURING DELIVERY feat(datapolicy): add recruitment data handling and backport changes

atm-quentin opened the pull request:
- Introduced new datapolicy entry for recruitment ("Candidature") with deletion delay configuration.
- Updated language files to include relevant labels.
- Enhanced setup page to dynamically render datapolicy configurations using a generic structure for better flexibility.
- Refactored cron task handling to accommodate recruitment-specific logic.
- Added support for the recruitment module in `datapolicy` and improved integration with existing system.

Closes #34704.

atm-quentin committed:
feat(datapolicy): add recruitment data handling and backport changes

- Introduced new datapolicy entry for recruitment ("Candidature") with deletion delay configuration.
- Updated language files to include relevant labels.
- Enhanced setup page to dynamically render datapolicy configurations using a generic structure for better flexibility.
- Refactored cron task handling to accommodate recruitment-specific logic.
- Added support for the recruitment module in `datapolicy` and improved integration with existing system.

Closes #34704.

atm-quentin committed:
FIX tms was empty

atm-jonathan commented on htdocs/datapolicy/admin/setup.php:
```diff
@@ -83,44 +204,38 @@
-				} else {
-					$val_const = GETPOSTINT($key);
+	// Loop through the data structure to find all possible constants to save.
+	foreach ($arrayofparameters as $title => $tab) {
```
Pas besoin du as $title ici ?

atm-quentin committed:
feat(extrafields): add hook points for enhanced customization

- Introduced multiple new hook points (`printFieldListTitle`, `printFieldListValue`, `formObjectOptions`, `loadExtrafieldsAttributes`, etc.) to allow enhanced customization and extension of extrafields functionalities.
- Updated templates and extrafields class to execute hooks for additional flexibility in rendering and data manipulation.
- Ensured backward compatibility with existing extrafields implementations.

Closes #34876.

atm-jonathan commented on htdocs/datapolicy/admin/setup.php:
```diff
@@ -132,78 +247,94 @@
+	$generateSelectHtml = function (string $constKey, array $valTab) {
+		$output = '<select name="'.$constKey.'" id="'.$constKey.'" class="flat minwidth200">';
+		foreach ($valTab as $key1 => $val1) {
+			$output .= '<option value="'.$key1.'" '.(getDolGlobalString($constKey) == $key1 ? 'selected="selected"' : '').'>';
```
tu peux utiliser dol_escape_htmltag() sur tes key et val ?

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+	 * Orchestrates the data cleaning process by iterating through all defined policies.
+	 * @return int Returns 0 for success, 1 for failure, as required for cron jobs.
+	 */
+	public function cleanDataForDataPolicy()
```
typage de retour ?

atm-quentin committed:
fix(extrafields): correct parameter passed to `executeHooks`

- Replaced `$object` with `$extrafields` as the parameter in `printFieldListTitle` hook execution to ensure proper data handling.
- Prevents potential issues where `$object` might not contain the required extrafields context.

No breaking changes.

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+		$this->db->begin();
+
+		// Iterate through each defined policy to apply its rules.
+		foreach ($dataPolicies as $policyKey => $policy) {
```
pas besoin de key

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+	 * @return void
 	 */
-	public function __construct(DoliDB $db)
+	private function _processPolicyAction($policy, $action, $object, &$processedIds, $conf, $user)
```
typage retour

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+	 * @param array<string, mixed> $policy The policy configuration.
+	 * @return int   The result of the delete operation.
+	 */
+	private function _handleDelete($object, $user, $policy)
```
typage ?

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+	 * @return int   The result of the update operation, or 0 if skipped.
 	 */
-	public function cleanDataForDataPolicy()
+	private function _handleAnonymize($object, $user, $policy)
```
idem

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
-					'state_id' => 1,
-					'socialnetworks' => [],
-					'country_id' => 0,
+		foreach ($policy['anonymize_fields'] as $field => $val) {
```
test si il est vide ?
il peut l'être :
ligne 371 : 'anonymize_fields' => array(),

atm-quentin replied:
c'est pas gênant car un foreach sur un tableau vide ne fait pas de warning, en général on met ce test si vraiment on a un string vide ou null / false par exemple

atm-jonathan commented on htdocs/datapolicy/class/datapolicycron.class.php:
```diff
@@ -29,503 +30,349 @@
+	 * @param string $action The action that was performed ('delete' or 'anonymize').
+	 * @return void
+	 */
+	private function _recordActionResult($result, $object, $action)
```
Typage ?

atm-jonathan commented on htdocs/core/tpl/admin_extrafields_edit.tpl.php:
```diff
@@ -316,7 +316,11 @@ function init_typeoffields(type)
 </td></tr>
-
+<?php
+$parameters = [];
```
array() ?

atm-jonathan commented on htdocs/datapolicy/admin/setup.php:
```diff
@@ -83,44 +204,38 @@
  * Actions
  */
+// Handle form submission to update constants
+if ($action == 'update' && empty($_POST['cancel'])) {
```
GETPOST

atm-quentin committed:
refactor(datapolicy): improve code readability and type safety

- Updated function and method definitions to include explicit return types for better type safety (`int`, `void`, `array`).
- Simplified loops by removing unused keys from `foreach` constructs.
- Minor adjustment in setup form, removing redundant cancel parameter value (`'cancel'` replaced with empty string).
- Improved maintainability of datapolicy cron class.

No breaking changes.

atm-quentin committed:
refactor(datapolicy): replace custom dropdown generation with `Form::selectarray`

- Removed the custom `$generateSelectHtml` function, simplifying code by using the existing `Form::selectarray` method.
- Updated dropdown generation for 'anonymize' and 'delete' actions to align with Dolibarr's standard practices.
- Enhances maintainability by leveraging built-in utilities and reducing custom logic.

Commit:
hook cdc interview

Commit:
Merge branch '20.0_cdcconseil_confidentialite' of github.com:ATM-Consulting/dolibarr into 20.0_cdcconseil_confidentialite

atm-quentin commented on htdocs/core/tpl/admin_extrafields_view.tpl.php:
```diff
@@ -92,6 +92,11 @@
 }
+$parameters = array();
+$hookmanager = new HookManager($db);
+$hookmanager->initHooks(array('extrafieldsdao'));
```
à déplacer dans ton setup normalement avec un inithook du style cdcinterview_setup, tu avais une fatal ?

atm-quentin commented on htdocs/core/tpl/admin_extrafields_view.tpl.php:
```diff
@@ -191,7 +199,7 @@
 	}
 } else {
-	$colspan = 17;
+	$colspan = 18;
```
modifie l'attribut en js dans ton setup

atm-quentin replied:
car là tu passe à 18 pour TOUT les setup extrafield qui utilise ce fichier

Commit:
delete init hook

Commit:
colspan
